EIPS/EIP-X-metadata-nickjohnson.md

```diff
@@ -0,0 +1,72 @@
+## Preamble
+
+    EIP: <to be assigned>
+    Title: Address metadata registry
+    Author: Nick Johnson <nick@ethereum.org>
+    Type: Standard track
+    Category: ERC
+    Status: Draft
+    Created: 2018-03-12
+
+## Abstract
+This EIP specifies a registry for address metadata, permitting both contracts and external accounts to supply metadata about themselves to onchain and offchain callers. This permits use-cases such as generalised authorisations, providing token acceptance settings, and claims registries.
+
+## Motivation
+An increasing set of use cases require storage of metadata associated with an address; see for instance EIP 777 and EIP 780, and the ENS reverse registry in EIP 181. Presently each use-case defines its own specialised registry. To prevent a proliferation of special-purpose registry contracts, we instead propose a single standardised registry using an extendable architecture that allows future standards to implement their own metadata standards.
+
+## Specification
+The metadata registry has the following interface:
+```
+interface AddressMetadataRegistry {
+  function provider(address target) view returns(address);
+  function setProvider(address _provider);
+}
+```
+
+`setProvider` specifies the metadata registry to be associated with the caller's address, while `provider` returns the address of the metadata registry for the supplied address.
+
+The metadata registry will be compiled with an agreed-upon version of Solidity and deployed using the trustless deployment mechanism to a fixed address that can be replicated across all chains.
+
+## Provider specification
+
+Providers may implement any subset of the metadata record types specified here. Where a record types specification requires a provider to provide multiple functions, the provider MUST implement either all or none of them. Providers MUST throw if called with an unsupported function ID.
+
+Providers have one mandatory function:
+
+```
+function supportsInterface(bytes4 interfaceID) constant returns (bool)
+```
+
+The `supportsInterface` function is documented in [EIP 165](https://github.com/ethereum/EIPs/issues/165), and returns true if the provider implements the interface specified by the provided 4 byte identifier. An interface identifier consists of the XOR of the function signature hashes of the functions provided by that interface; in the degenerate case of single-function interfaces, it is simply equal to the signature hash of that function. If a provider returns `true` for `supportsInterface()`, it must implement the functions specified in that interface.
```

Comment:
What is the motivation behind using XOR of the signature IDs rather than a hash of all of the signatures?  While I suspect someone can implement the XOR without too much difficulty, keccak hashing of things in the Solidity ecosystem is a well understood problem as it is used everywhere while XOR of first 4 bytes of a keccak hash of a thing is significantly less understood by the average developer.

The following is what the hash might look like, which is pretty familiar to any Ethereum developer at this point and can be done in Solidity or in JavaScript (most common dapp development language).
```
keccak256("function foo(uint256,bool)", "function bar()", "function zip(string,bytes)")
```

Reply:
See EIP 165 for justification. I don't think this is a good place to rehash that.

Reply:
Glanced through that, there was a reference to a Gitter conversation where `bytes4` XOR was chosen over bytes32 hash but no details of the contents of that conversation were provided.  From my reading of #165, I am not at all convinced that XOR of function IDs is superior to hash of function signatures.

Was there a particular comment that you feel captures the argument over there?  It is possible I missed a subtle argument as I was skimming the comments rather than analyzing them in depth.

Reply:
I really don't want to get derailed on arguing the merits of an unrelated standard in this EIP. Can you open a discussion on 165 instead? It's unlikely we can make a change at this point, though, since it's already widely deployed (eg, on ENS for one).

Reply:
Yeah, now that I have read EIP 165 and realize that this section is just a watered down summary of 165 I no longer want to press the bytes32 vs XOR thing here.